### ISSUE TITLE
Wrong type argument: stringp, :json-false when trying to load pinboard

### ISSUE THREAD

**Issue by baron** [davep/pinboard.el#7]
I get "Wrong type argument: stringp, :json-false" when trying to call "pinboard". Not sure what's going on.

**Comment by baron** [comment]
It turns out I had several blank descriptions which caused parsing errors as they default to ":json-false".

```elisp
(defun pinboard-call (url caller)
  "Call on URL and return the data.

CALLER is a symbol that is the name of the caller. This is used
to help set rate limits."
  (let ((url-request-extra-headers `(("User-Agent" . ,pinboard-agent)))
        (url-show-status nil))
    (pinboard-remember-call caller)
    (with-temp-buffer
      (url-insert-file-contents url)
      (condition-case err
          (let ((json-false ""))
            (json-read))
        (error
         (error "Error '%s' handling reply from Pinboard: %s"
                (error-message-string err) (buffer-string)))))))
```

Thank you for this library.

**Comment by davep** [comment]
Could you give me an example of how to recreate this? I created a pin with everything blank apart from the URL itself:

![image](https://user-images.githubusercontent.com/28237/60003141-ff41b080-9661-11e9-9f19-0e97aa0579ea.png)

and I can't get the same error.

Tested with Emacs 26.2 (and so `json.el` v1.4).

**Comment by davep** [comment]
Just testing a little further, if I create a pin like the above, and then pull down the JSON of it, it looks like this:

```json
{
"href":"http:\/\/www.example.com\/",
"description":"",
"extended":"",
"meta":"4f0fd3af112ff936bf06cab30c7604cf",
"hash":"f1777111f5d0f1c81ffa04de751128fa",
"time":"2019-06-24T08:52:38Z",
"shared":"no",
"toread":"no",
"tags":""
}
```

If I paste that into a buffer and then <kbd>M-:</kbd> <kbd>(json-read)</kbd> <kbd>RET</kbd> it, it reads just fine.

**Comment by baron** [comment]
Thanks for the follow up.

What is the value of "json-false" in your environment?

If you could restart emacs by setqing the values of json-false and json-nil to the default values and restarting emacs you might be able to reproduce the bug.

```
(defvar json-false :json-false
  "Value to use when reading JSON `false'.
If this has the same value as `json-null', you might not be able to tell
the difference between `false' and `null'.  Consider let-binding this
around your call to `json-read' instead of `setq'ing it.")

(defvar json-null nil
  "Value to use when reading JSON `null'.
If this has the same value as `json-false', you might not be able to
tell the difference between `false' and `null'.  Consider let-binding
this around your call to `json-read' instead of `setq'ing it.")
```

I have been able to reproduce the bug with a complete vanilla emacs (on osx) with a fresh init file with only package, melpa, and pinboard  respectively.

![スクリーンショット 2019-06-25 13 04 24](https://user-images.githubusercontent.com/2617/60068569-4905c600-974a-11e9-9c45-534f446e1040.png)

In theory your data should trigger this bug if json-false is also set to ":json-false".

**Comment by baron** [comment]
>                                                                1970-01-01 09:00:00  /
>                                                                1970-01-01 09:00:00  http://twitter.com/pocket

It might be unrelated but I seem to have some weird data in my bookmarks that could be the culprit.

**Comment by davep** [comment]
`json-null` is `nil` and `json-false` is `:json-false` in my environment.

> In theory your data should trigger this bug if json-false is also set to ":json-false".

The idea being that an empty string would come out as `:json-false`? I'm not seeing that at all:

```elisp
*** Welcome to IELM ***  Type (describe-mode) for help.
ELISP> json-false
:json-false
ELISP> json-null
nil
ELISP> (json-read-from-string "{\"foo\":\"\"}")
((foo . ""))
```

If you do the [parse test I mention above](https://github.com/davep/pinboard.el/issues/7#issuecomment-504921841), what do you see? It seems odd to me that an empty string would be parsed as any form of `false` value.

Are you able to isolate the json of a problematic pin so we can see what Pinboard is sending you?

**Comment by davep** [comment]
Did you get a chance to do the above test @baron?

**Comment by davep** [comment]
Given no more feedback, and no other reports of this, I'm going to close this.

**Comment by travisjeffery** [comment]
I've got the same thing. 

Stacktrace:

```
Debugger entered--Lisp error: (wrong-type-argument stringp :json-false)
  propertize(:json-false font-lock-face pinboard-read-face)
  #f(compiled-function (s pin) #<bytecode 0x15726f97f829>)(:json-false ((href . "http://www.nytimes.com/2010/03/21/opinion/21Benkle...") (description . :json-false) (extended . "Why $33/mo in France buys broadband internet, hi-d...") (meta . "6648db09c90261f24497edb06f42f11b") (hash . "0b6dfbaf771d7becc5f195ded423d09a") (time . "2010-03-21T16:39:59Z") (shared . "yes") (toread . "no") (tags . "")))
  #f(compiled-function (pin) #<bytecode 0x157270d0dfc5>)(((href . "http://www.nytimes.com/2010/03/21/opinion/21Benkle...") (description . :json-false) (extended . "Why $33/mo in France buys broadband internet, hi-d...") (meta . "6648db09c90261f24497edb06f42f11b") (hash . "0b6dfbaf771d7becc5f195ded423d09a") (time . "2010-03-21T16:39:59Z") (shared . "yes") (toread . "no") (tags . "")))
  mapcar(#f(compiled-function (pin) #<bytecode 0x157270d0dfc5>) ((... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) (... ... ... ... ... ... ... ... ...) ...))
  pinboard-redraw()
  pinboard-refresh()
  funcall-interactively(pinboard-refresh)
  #<subr call-interactively>(pinboard-refresh nil nil)
  apply(#<subr call-interactively> pinboard-refresh (nil nil))
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> pinboard-refresh nil nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (pinboard-refresh nil nil))
  call-interactively(pinboard-refresh nil nil)
  command-execute(pinboard-refresh)
```

json-false:

```
json-false is a variable defined in ‘json.el’.
Its value is ‘:json-false’
```

Emacs version:

```
GNU Emacs 27.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.22.30) of 2019-08-29
```

**Comment by travisjeffery** [comment]
I think it might be due to really old bookmarks that were nullable/null by default and now pinboard defaults them to "".

**Comment by travisjeffery** [comment]
@davep ^

**Comment by davep** [comment]
Thanks for that. I’ll take a look in the next day or so. Do you have an example of the sort of problematic JSON you talk about?

**Comment by baron** [comment]
Sorry I couldn't help with the feedback. As noted above I solved the problem by binding json-false to a blank string ("") before parsing.

As you can see in @travisjeffery's error report, blank strings become ":json-false" which throws an error.

```(description . :json-false)```

The source to json.el says:

```
(defvar json-false :json-false
  "Value to use when reading JSON `false'.
If this has the same value as `json-null', you might not be able to tell
the difference between `false' and `null'.  Consider let-binding this
around your call to `json-read' instead of `setq'ing it.")
```

And that's what I did.


@travisjeffery I've updated the code below. Eval it in a scratch buffer and see if the problem goes away.

```
  (defun pinboard-call (url caller)
    "Call on URL and return the data.

CALLER is a symbol that is the name of the caller. This is used
to help set rate limits."
    (let ((url-request-extra-headers `(("User-Agent" . ,pinboard-agent)))
          (url-show-status nil))
      (pinboard-remember-call caller)
      (with-temp-buffer
        (url-insert-file-contents url)
        (condition-case err
            (let ((json-false ""))
              (json-read-from-string (buffer-string)))
          (error
           (error "Error '%s' handling reply from Pinboard: %s"
                  (error-message-string err) (buffer-string)))))))
```

**Comment by davep** [comment]
Hey @baron, thanks for that. I don't think I'd noticed that at the time and, bizarrely, I was just tinkering with exactly that workaround as you commented!

But, yeah, you're right, that seems to be the way to go. Thanks. :)

**Comment by baron** [comment]
Great to hear that! I look forward to retiring my weird hot patch.

**Comment by davep** [comment]
Fixed applied in fc967e1583312ef727785d236519c7a09bf92be3 -- thanks both for your input on this.
